File: task1solved/src/EventBus.ts
import Vue from "vue";
import INotification from "@/domain/notification";

export enum EventBusEvents {
    DISPLAY_NOTIFICATION = "display-notification"
}

export interface NotificationEventHandler {
    (notification: INotification): void;
}

export interface IEventBus {
    $on(event: EventBusEvents.DISPLAY_NOTIFICATION, callback: NotificationEventHandler): this;
    $on(event: EventBusEvents, callback: Function): this;

    $once(event: EventBusEvents.DISPLAY_NOTIFICATION, callback: NotificationEventHandler): this;
    $once(event: EventBusEvents, callback: Function): this;

    $off(event: EventBusEvents.DISPLAY_NOTIFICATION, callback: NotificationEventHandler): this;
    $off(event?: EventBusEvents, callback?: Function): this;

    $emit(event: EventBusEvents.DISPLAY_NOTIFICATION, notification: INotification): this;
    $emit(event: EventBusEvents, ...args: any[]): this;
}

export const EventBus: IEventBus = new Vue();

File: spa-solved-blazorwasm-port/EventBus.cs
namespace Forse.Shop;

public static class EventBusEvents
{
    public const string DisplayNotification = "display-notification";
}

public class EventBus
{
    private readonly Dictionary<string, List<EventData>> _eventTable = new();

    public void On(string eventName, Action<object> listener)
    {
        if (!_eventTable.ContainsKey(eventName))
            _eventTable.Add(eventName, new List<EventData>());

        var data = new EventData(eventName, listener, false);
        
        _eventTable[eventName].Add(data);
    }

    public void Once(string eventName, Action<object> listener)
    {
        if (!_eventTable.ContainsKey(eventName))
            _eventTable.Add(eventName, new List<EventData>());

        var data = new EventData(eventName, listener, true);
        
        _eventTable[eventName].Add(data);
    }

    public void Off(string eventName)
    {
        if (!_eventTable.ContainsKey(eventName)) 
            return;

        _eventTable[eventName].Clear();
    }
    
    public void Off(string eventName, Action<object> listener)
    {
        if (!_eventTable.ContainsKey(eventName)) 
            return;

        var data = _eventTable[eventName].FirstOrDefault(data => data.Listener == listener);
        if (data == null)
            return;
        
        _eventTable[eventName].Remove(data);
    }

    public void Emit(string eventName, object data)
    {
        if (!_eventTable.ContainsKey(eventName)) 
            return;

        foreach (var eventData in _eventTable[eventName])
            eventData.Listener(data);

        _eventTable[eventName].RemoveAll(eventData => eventData.IsOnce);
    }

    private record EventData(string EventName, Action<object> Listener, bool IsOnce);
}    
